File: 13.SlidingWindow滑动窗口/leetcode_1358_numberOfSubstrings.py
from collections import defaultdict

class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        r = 0
        cnt = defaultdict(int)
        res = 0
        for l in range(n):
            while r < n and (cnt['a'] < 1 or cnt['b'] < 1 or cnt['c'] < 1):
                cnt[s[r]] += 1
                r += 1
            if cnt['a'] >= 1 and cnt['b'] >= 1 and cnt['c'] >= 1:
                res += n-r+1
            cnt[s[l]] -= 1
        return res

if __name__ == "__main__":
    sl = Solution()
    s = "abcabc"
    print(sl.numberOfSubstrings(s))

File: 13.SlidingWindow滑动窗口/leetcode_1358_numberOfSubstrings.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int numberOfSubstrings(string s) {
        int n = s.size();
        int l = 0;
        unordered_map<char, int> cnt;
        int res = 0;
        for (int r=0; r<n; r++){
            cnt[s[r]]++;
            while (cnt['a'] >= 1 && cnt['b'] >= 1 && cnt['c'] >= 1){
                cnt[s[l]]--;
                l++;
            }
            res += l;
        }
        return res;
    }
};

int main(){
    Solution sl;
    string s = "abcabc";
    cout << sl.numberOfSubstrings(s) << endl;
}
